File: 0547-number-of-provinces/0547-number-of-provinces.java
class Solution {
    int n;

    public void dfs(int i, List<List<Integer>> adj, boolean[] vis){
        vis[i] = true;

        for(int neigh : adj.get(i)){
            if(!vis[neigh]){
                dfs(neigh, adj, vis);
            }
        }
    }

    public int findCircleNum(int[][] isConnected) {
        this.n = isConnected.length;
        List<List<Integer>> adj = new ArrayList<>();

        for(int i=0; i<n; i++){
            adj.add(new ArrayList<>());
        }

        for(int i=0; i<n; i++){
            for(int j=0; j<n; j++){
                if(isConnected[i][j]==1){
                    adj.get(i).add(j);
                    adj.get(j).add(i);
                }
            }
        }

        boolean[] visited = new boolean[n];
        int res = 0;

        for(int i=0; i<n; i++){
            if(!visited[i]){
                res++;
                dfs(i, adj, visited);
            }
        }

        return res;
    }
}

File: 0547-number-of-provinces/0547-number-of-provinces.py
class Solution(object):
    def findCircleNum(self, isConnected):
        """
        :type isConnected: List[List[int]]
        :rtype: int
        """
        m = len(isConnected)
        n = len(isConnected[0])

        adj = defaultdict(list)

        for i in range(m):
            for j in range(n):
                if isConnected[i][j] == 1:
                    adj[i].append(j)
                    adj[j].append(i)

        nodes = len(adj)
        visited = [False for _ in range(nodes)]

        def dfs(node):
            visited[node] = True

            for nxt in adj[node]:
                if not visited[nxt]:
                    dfs(nxt)

        count = 0
        for key in adj:
            if not visited[key]:
                dfs(key)
                count+=1

        return count 
